lightning/mydump/region.go

```diff
@@ -133,77 +137,140 @@ func MakeTableRegions(
 	store storage.ExternalStorage,
 ) ([]*TableRegion, error) {
 	// Split files into regions
-	filesRegions := make([]*TableRegion, 0, len(meta.DataFiles))
-	dataFileSizes := make([]float64, 0, len(meta.DataFiles))
-	prevRowIDMax := int64(0)
-	var err error
-	for _, dataFile := range meta.DataFiles {
-		if dataFile.FileMeta.Type == SourceTypeParquet {
-			rowIDMax, region, err := makeParquetFileRegion(ctx, store, meta, dataFile, prevRowIDMax)
-			if err != nil {
-				return nil, err
-			}
-			prevRowIDMax = rowIDMax
-			filesRegions = append(filesRegions, region)
-			// TODO: how to estimate raw data size for parquet file
-			dataFileSizes = append(dataFileSizes, float64(dataFile.Size))
-			continue
-		}
+	type fileRegionRes struct {
+		info    FileInfo
+		regions []*TableRegion
+		sizes   []float64
+		err     error
+	}
 
-		dataFileSize := dataFile.Size
+	start := time.Now()
 
-		divisor := int64(columns)
-		isCsvFile := dataFile.FileMeta.Type == SourceTypeCSV
-		if !isCsvFile {
-			divisor += 2
-		}
+	execCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-		// If a csv file is overlarge, we need to split it into multiple regions.
-		// Note: We can only split a csv file whose format is strict.
-		if isCsvFile && dataFileSize > int64(cfg.Mydumper.MaxRegionSize) && cfg.Mydumper.StrictFormat {
-			var (
-				regions      []*TableRegion
-				subFileSizes []float64
-			)
-			prevRowIDMax, regions, subFileSizes, err = SplitLargeFile(ctx, meta, cfg, dataFile, divisor, prevRowIDMax, ioWorkers, store)
-			if err != nil {
-				return nil, err
+	concurrency := utils.MaxInt(cfg.App.RegionConcurrency, 2)
+	fileChan := make(chan FileInfo, concurrency)
+	resultChan := make(chan fileRegionRes, concurrency)
+	var wg sync.WaitGroup
+	for i := 0; i < concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			for info := range fileChan {
+				regions, sizes, err := makeSourceFileRegion(execCtx, meta, info, columns, cfg, ioWorkers, store)
+				resultChan <- fileRegionRes{info: info, regions: regions, sizes: sizes, err: err}
 			}
-			dataFileSizes = append(dataFileSizes, subFileSizes...)
-			filesRegions = append(filesRegions, regions...)
-			continue
+			wg.Done()
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		close(resultChan)
+	}()
+
+	errChan := make(chan error, 1)
+	fileRegionsMap := make(map[string]fileRegionRes, len(meta.DataFiles))
+	go func() {
+		for res := range resultChan {
+			if res.err != nil {
+				errChan <- res.err
+				return
+			}
+			fileRegionsMap[res.info.FileMeta.Path] = res
 		}
-		rowIDMax := prevRowIDMax + dataFile.Size/divisor
-		tableRegion := &TableRegion{
-			DB:       meta.DB,
-			Table:    meta.Name,
-			FileMeta: dataFile.FileMeta,
-			Chunk: Chunk{
-				Offset:       0,
-				EndOffset:    dataFile.Size,
-				PrevRowIDMax: prevRowIDMax,
-				RowIDMax:     rowIDMax,
-			},
+		errChan <- nil
+	}()
+
+	for _, dataFile := range meta.DataFiles {
+		fileChan <- dataFile
+	}
+	close(fileChan)
```

Comment:
move this close below `err := <-errChan`? 🤔 above goroutine in line 159 is reading from it